### PR TITLE
conn_pool_grid: try TCP immediately with h3 state FailedRecently

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: Add a new state FailedRecently to the status tracker. And make the connectivity grid check this state in newStream(). If the h3 pool is in this state, do not delay TCP racing.

Risk Level: low, grid only
Testing: added unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Part of #18715

### CODE REVIEW

**Comment by danzh2010** [comment]
failed thrift_proxy:integration_test on Windows seems unrelated.

File: source/common/http/conn_pool_grid.cc [hunk 1]
```diff
@@ -299,7 +302,14 @@ ConnectionPool::Cancellable* ConnectivityGrid::newStream(Http::ResponseDecoder&
     createNextPool();
   }
   PoolIterator pool = pools_.begin();
-  if (!shouldAttemptHttp3() || !options.can_use_http3_) {
+  Instance::StreamOptions overriding_options(options);
+  bool delay_tcp_attempt{true};
```
Comment (RyanTheOptimist) [inline source/common/http/conn_pool_grid.cc:306]: nit: we have two initializations, one uses {} and the other (). Could they both use {} (or even = to be more idiomatic)?

File: test/common/http/conn_pool_grid_test.cc [hunk 2]
```diff
@@ -837,6 +845,69 @@ TEST_F(ConnectivityGridTest, SuccessWithoutHttp3NoMatchingAlpn) {
...
+  EXPECT_CALL(*failover_timer, enabled()).WillOnce(Return(false)).WillOnce(Return(true));
+  EXPECT_CALL(*failover_timer, enableTimer(_, _));
+  EXPECT_NE(grid_->newStream(decoder_, callbacks_,
+                             {/*can_send_early_data_=*/true,
+                              /*can_use_http3_=*/true}),
+            nullptr);
+  EXPECT_NE(grid_->first(), nullptr);
+  // The 2nd pool should be TCP pool and it should have been created together with h3 pool.
+  EXPECT_NE(grid_->second(), nullptr);
+  EXPECT_EQ(2u, grid_->callbacks_.size());
```
Comment (RyanTheOptimist) [inline test/common/http/conn_pool_grid_test.cc:867]: I'm not quite sure I understand how this test that the TCP attempt is not delayed?

File: source/common/http/conn_pool_grid.cc [hunk 3]
```diff
@@ -317,6 +327,10 @@ ConnectionPool::Cancellable* ConnectivityGrid::newStream(Http::ResponseDecoder&
     // WrappedCallbacks object has also been deleted.
     return nullptr;
   }
+  if (!delay_tcp_attempt) {
+    // Immediately start TCP attempt if HTTP/3 failed recently.
+    wrapped_callbacks_.front()->tryAnotherConnection();
```
Comment (RyanTheOptimist) [inline source/common/http/conn_pool_grid.cc:332]: Should we stop the failover timer at this point since there is nothing else to try?

**Review by RyanTheOptimist — COMMENTED** [review]
Looks great!

File: test/common/http/conn_pool_grid_test.cc (hunk 2 above)
Comment (RyanTheOptimist) [reply test/common/http/conn_pool_grid_test.cc:867]: Ah, thanks. I see.

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
FYI: I've enabled auto-merge so once CI passes it should land automatically.

**Pull request merged**